interpreter/ruby/ruby.go

```diff
@@ -740,26 +776,81 @@ func Loader(ebpf interpreter.EbpfHandler, info *interpreter.LoaderInfo) (interpr
 	if version < rubyVersion(3, 0, 0) {
 		currentCtxSymbol = "ruby_current_execution_context_ptr"
 	}
-	currentCtxPtr, err := ef.LookupSymbolAddress(currentCtxSymbol)
-	if err != nil {
-		return nil, fmt.Errorf("%v not found: %v", currentCtxSymbol, err)
-	}
+
+	var currentEcTpBaseTlsOffset libpf.Address
+	var interpRanges []util.Range
 
 	// rb_vm_exec is used to execute the Ruby frames in the Ruby VM and is called within
 	// ruby_run_node  which is the main executor function since Ruby v1.9.0
 	// https://github.com/ruby/ruby/blob/587e6800086764a1b7c959976acef33e230dccc2/main.c#L47
-	symbolName := libpf.SymbolName("rb_vm_exec")
+	interpSymbolName := libpf.SymbolName("rb_vm_exec")
 	if version < rubyVersion(2, 6, 0) {
-		symbolName = libpf.SymbolName("ruby_exec_node")
+		interpSymbolName = libpf.SymbolName("ruby_exec_node")
 	}
-	interpRanges, err := info.GetSymbolAsRanges(symbolName)
+
+	var currentEcSymbol *libpf.Symbol
+	currentEcSymbolName := libpf.SymbolName(rubyCurrentEcTlsSymbol)
+
+	log.Debugf("Ruby %d.%d.%d detected, looking for currentCtxPtr=%q, currentEcSymbol=%q",
+		(version>>16)&0xff, (version>>8)&0xff, version&0xff, currentCtxSymbol, currentEcSymbolName)
+
+	// Symbol discovery strategy:
+	// - Ruby < 3.0.4: Uses currentCtxPtr (global/ractor-based execution context)
+	// - Ruby >= 3.0.4: Uses currentEcSymbol (TLS-based execution context via ruby_current_ec)
+	// When direct lookup fails, VisitSymbols scans all symbols as fallback.
+	// eBPF selects the appropriate method based on version at runtime.
+	currentCtxPtr, err := ef.LookupSymbolAddress(currentCtxSymbol)
 	if err != nil {
-		return nil, err
+		log.Debugf("Direct lookup of %v failed: %v, will try fallback", currentCtxSymbol, err)
+	}
+
+	interpRanges, err = info.GetSymbolAsRanges(interpSymbolName)
+	if err != nil {
+		log.Debugf("Direct lookup of %v failed: %v, will try fallback", interpSymbolName, err)
+	}
+
+	if err = ef.VisitSymbols(func(s libpf.Symbol) bool {
+		if s.Name == currentEcSymbolName {
+			currentEcSymbol = &s
+		}
+		if s.Name == currentCtxSymbol {
+			currentCtxPtr = s.Address
+		}
+		if len(interpRanges) == 0 && s.Name == interpSymbolName {
+			interpRanges = []util.Range{{
+				Start: uint64(s.Address),
+				End:   uint64(s.Address) + s.Size,
+			}}
+		}
+		if len(interpRanges) > 0 && currentEcSymbol != nil && currentCtxPtr != 0 {
+			return false
+		}
+		return true
+	}); err != nil {
+		log.Warnf("failed to visit symbols: %v", err)
+	}
+
+	// NOTE for ruby 3.3.0+, if ruby is stripped, we have no way of locating
+	// ruby_current_ec TLS symbol.
+	// We could potentially add a fallback for this in the future, but for now
+	// only unstripped ruby is supported. Many distro supplied rubies are stripped.
+	if err = ef.VisitTLSRelocations(func(r pfelf.ElfReloc, symName string) bool {
+		if symName == rubyCurrentEcTlsSymbol ||
+			currentEcSymbol != nil && libpf.SymbolValue(r.Addend) == currentEcSymbol.Address {
```

Comment:
Could `currentEcSymbol` be converted to `libpf.SymbolValue` to which the address is resolved at the earlier visitor. Would get rid of the `nil` compare portion here.